src/megatron/energon/watchdog.py

```diff
@@ -0,0 +1,291 @@
+        """
+        self._timeout = timeout
+        self._callback = callback
+        self._dump_stacks = dump_stacks
+
+        # If _deadline is None, the watchdog is disabled.
+        # Otherwise, _deadline = time.time() + _timeout if enabled.
+        self._deadline: Optional[float] = perf_counter() + timeout if enabled else None
+
+        self._stop = False  # signals permanent shutdown (finish)
+
+        # Condition variable to manage state changes
+        self._cv = threading.Condition()
+        # Background thread (daemon) that monitors timeouts
+        self._worker_thread = threading.Thread(target=self._worker, daemon=True)
+        self._worker_thread.start()
+
+    def _worker(self) -> None:
+        """
+        Background thread that periodically checks if the watchdog has expired.
+        Once it times out or is told to stop, it exits.
+        """
+        while True:
+            with self._cv:
+                if self._stop:
+                    # finish() was called; end the worker.
+                    return
+
+                if self._deadline is None:
+                    # Disabled; no deadline. Just wait a bit, then re-check.
+                    self._cv.wait(timeout=1.0)
+                    continue
+
+                remaining = self._deadline - perf_counter()
+                if remaining <= 0:
+                    # We have timed out
+                    self._trigger()
+                    return
+                else:
+                    # Wait until either the deadline or a state change
+                    self._cv.wait(timeout=remaining)
+
+    def _trigger(self) -> None:
+        """
+        Called exactly once if the watchdog times out.
+        1) Optionally dumps stacks,
+        2) Calls user callback,
+        3) If callback raises an error,
+           - print traceback,
+           - sys.exit(1),
+           - fallback to os._exit(1) after 10s if process not terminated.
+        """
+        watchdog_thread_id = threading.get_ident()
+
+        # 1) Dump stacks if requested
+        if self._dump_stacks:
+            print("Watchdog triggered: Dumping thread stacks")
+            self._print_all_thread_stacks(skip_thread_id=watchdog_thread_id)
+
+        # 2) Call user callback
+        if self._callback:
+            try:
+                self._callback()
+            except Exception:
+                # Print the traceback
+                traceback.print_exc()
+
+                # Start a background kill-switch after 10 seconds
+                def force_exit_after_delay() -> None:
+                    time.sleep(10)
+                    os._exit(1)
+
+                killer = threading.Thread(target=force_exit_after_delay, daemon=True)
+                killer.start()
+
+                # Attempt graceful shutdown
+                sys.exit(1)
+
+    def _print_all_thread_stacks(self, skip_thread_id: Optional[int] = None) -> None:
```

Comment:
Shouldn't this even be limited to the thread, that created the Watchdog? (maybe with option)